### ISSUE TITLE
x86/avx512 vperm optimizations can produce incorrect permutations

### ISSUE THREAD

**Issue by jlainema** [llvm/llvm-project#127054]
// this code should not report mismatches when run on a cpu with AVX512F support:
```c
#include <cstdint>
#include <cstdlib>
#include <cstdio>
#include <immintrin.h>

alignas(64) static const __m512i swizzles_512[10] = {
  _mm512_set_epi32( 0, 1, 2, 3, 4, 5, 6, 7,16,17,18,19,20,21,22,23),
  _mm512_set_epi32( 8, 9,10,11,12,13,14,15,24,25,26,27,28,29,30,31),
  _mm512_set_epi32( 0, 1, 2, 3,16,17,18,19, 8, 9,10,11,24,25,26,27),
  _mm512_set_epi32( 4, 5, 6, 7,20,21,22,23,12,13,14,15,28,29,30,31),
  _mm512_set_epi32( 0, 1,16,17, 4, 5,20,21, 8, 9,24,25,12,13,28,29),
  _mm512_set_epi32( 2, 3,18,19, 6, 7,22,23,10,11,26,27,14,15,30,31),
  _mm512_set_epi32( 0,16, 2,18, 4,20, 6,22, 8,24,10,26,12,28,14,30),
  _mm512_set_epi32( 1,17, 3,19, 5,21, 7,23, 9,25,11,27,13,29,15,31),
  _mm512_set_epi32( 0,16, 1,17, 2,18, 3,19, 4,20, 5,21, 6,22, 7,23),
  _mm512_set_epi32( 8,24, 9,25,10,26,11,27,12,28,13,29,14,30,15,31)
};

alignas(64) uint32_t swizzles[10][16] = {
  { 0, 1, 2, 3, 4, 5, 6, 7,16,17,18,19,20,21,22,23},{ 8, 9,10,11,12,13,14,15,24,25,26,27,28,29,30,31},
  { 0, 1, 2, 3,16,17,18,19, 8, 9,10,11,24,25,26,27},{ 4, 5, 6, 7,20,21,22,23,12,13,14,15,28,29,30,31},
  { 0, 1,16,17, 4, 5,20,21, 8, 9,24,25,12,13,28,29},{ 2, 3,18,19, 6, 7,22,23,10,11,26,27,14,15,30,31},
  { 0,16, 2,18, 4,20, 6,22, 8,24,10,26,12,28,14,30},{ 1,17, 3,19, 5,21, 7,23, 9,25,11,27,13,29,15,31},
  { 0,16, 1,17, 2,18, 3,19, 4,20, 5,21, 6,22, 7,23},{ 8,24, 9,25,10,26,11,27,12,28,13,29,14,30,15,31}};

void reorder_gather512(uint32_t *_a) {
  __m512i a,b;
  a = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[0]), _a, 4);  
  b = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[1]), _a, 4);
  _mm512_store_epi32(_a,_mm512_min_epi32(a,b));
  _mm512_store_epi32(_a+16,_mm512_max_epi32(a,b));
  a = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[2]), _a, 4);  
  b = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[3]), _a, 4);
  _mm512_store_epi32(_a,_mm512_min_epi32(a,b));
  _mm512_store_epi32(_a+16,_mm512_max_epi32(a,b));
  a = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[4]), _a, 4);  
  b = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[5]), _a, 4);
  _mm512_store_epi32(_a,_mm512_min_epi32(a,b));
  _mm512_store_epi32(_a+16,_mm512_max_epi32(a,b));
  a = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[6]), _a, 4);  
  b = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[7]), _a, 4);
  _mm512_store_epi32(_a,_mm512_min_epi32(a,b));
  _mm512_store_epi32(_a+16,_mm512_max_epi32(a,b));
  a = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[8]), _a, 4);  
  b = _mm512_i32gather_epi32(_mm512_load_epi32(swizzles[9]), _a, 4);
  _mm512_store_epi32(_a,a);
  _mm512_store_epi32(_a+16,b);
}

#define PERM_ACCESS(n) swizzles_512[n]
// #define PERM_ACCESS(n) _mm512_load_epi32(swizzles[n])

void reorder_perm512(uint32_t *_a) {
  __m512i a = _mm512_load_epi32(_a), b = _mm512_load_epi32(_a+16);
  __m512i x,y;
  x = _mm512_permutex2var_epi32(a,PERM_ACCESS(0),b);
  y = _mm512_permutex2var_epi32(a,PERM_ACCESS(1),b);
  a = _mm512_min_epi32(x,y);
  b = _mm512_max_epi32(x,y);
  x = _mm512_permutex2var_epi32(a,PERM_ACCESS(2),b);
  y = _mm512_permutex2var_epi32(a,PERM_ACCESS(3),b);
  a = _mm512_min_epi32(x,y);
  b = _mm512_max_epi32(x,y);
  x = _mm512_permutex2var_epi32(a,PERM_ACCESS(4),b);
  y = _mm512_permutex2var_epi32(a,PERM_ACCESS(5),b);
  a = _mm512_min_epi32(x,y);
  b = _mm512_max_epi32(x,y);
  x = _mm512_permutex2var_epi32(a,PERM_ACCESS(6),b);
  y = _mm512_permutex2var_epi32(a,PERM_ACCESS(7),b);
  a = _mm512_min_epi32(x,y);
  b = _mm512_max_epi32(x,y);
  _mm512_store_epi32(_a,_mm512_permutex2var_epi32(a,PERM_ACCESS(8),b));
  _mm512_store_epi32(_a+16,_mm512_permutex2var_epi32(a,PERM_ACCESS(9),b));
}

int main() {
  alignas(64) uint32_t aval[32], bval[32];

  for(unsigned i=0; i<32; i++) aval[i] = bval[i] = rand();

  reorder_perm512(aval);
  reorder_gather512(bval);

  for (unsigned i=0; i<10; i++) {
    __m512i dx = _mm512_sub_epi32(swizzles_512[i], _mm512_load_epi32(swizzles[i]));
    if (_mm512_reduce_add_epi32(dx))
      printf("index mismatch for register %d\n", i);
  }

  for (unsigned i=0; i<32; i++) {
    if (aval[i] != bval[i]) {
      printf("gather/permute mismatch %08x != %08x (@%d)\n", aval[i], bval[i], i);
    }
  }

  return 0;
}
```

**Comment by jlainema** [comment]
Same behavior on clang/llvm version 18 on Ubuntu 24.04 and version 21 built from git cloned yesterday.

**Comment by RKSimon** [comment]
Testing on gcc and clang I get the same mismatches:
```
simon:~/LLVM$ gcc -g0 -O3 -march=native pr127054.cpp
simon:~/LLVM$ ./a.out
gather/permute mismatch 327b23c6 != 19495cff (@0)
gather/permute mismatch 2eb141f2 != 238e1f29 (@1)
gather/permute mismatch 625558ec != 2ae8944a (@2)
gather/permute mismatch 3d1b58ba != 507ed7ab (@3)
gather/permute mismatch 238e1f29 != 2eb141f2 (@4)
gather/permute mismatch 19495cff != 327b23c6 (@5)
gather/permute mismatch 507ed7ab != 3d1b58ba (@6)
gather/permute mismatch 2ae8944a != 625558ec (@7)
gather/permute mismatch 74b0dc51 != 41b71efb (@8)
gather/permute mismatch 46e87ccd != 643c9869 (@9)
gather/permute mismatch 79e2a9e3 != 66334873 (@10)
gather/permute mismatch 7545e146 != 6b8b4567 (@11)
gather/permute mismatch 643c9869 != 46e87ccd (@12)
gather/permute mismatch 41b71efb != 74b0dc51 (@13)
gather/permute mismatch 6b8b4567 != 7545e146 (@14)
gather/permute mismatch 66334873 != 79e2a9e3 (@15)
gather/permute mismatch 3352255a != 0216231b (@16)
gather/permute mismatch 1190cde7 != 0ded7263 (@17)
gather/permute mismatch 66ef438d != 109cf92e (@18)
gather/permute mismatch 140e0f76 != 1befd79f (@19)
gather/permute mismatch 0ded7263 != 1190cde7 (@20)
gather/permute mismatch 0216231b != 3352255a (@21)
gather/permute mismatch 1befd79f != 140e0f76 (@22)
gather/permute mismatch 109cf92e != 66ef438d (@23)
gather/permute mismatch 5bd062c2 != 12200854 (@24)
gather/permute mismatch 41a7c4c9 != 1f16e9e8 (@25)
gather/permute mismatch 7fdcc233 != 4db127f8 (@26)
gather/permute mismatch 6b68079a != 515f007c (@27)
gather/permute mismatch 1f16e9e8 != 41a7c4c9 (@28)
gather/permute mismatch 12200854 != 5bd062c2 (@29)
gather/permute mismatch 515f007c != 6b68079a (@30)
gather/permute mismatch 4db127f8 != 7fdcc233 (@31)

simon:~/LLVM$ clang -g0 -O3 -march=native pr127054.cpp
simon:~/LLVM$ ./a.out
gather/permute mismatch 327b23c6 != 19495cff (@0)
gather/permute mismatch 2eb141f2 != 238e1f29 (@1)
gather/permute mismatch 625558ec != 2ae8944a (@2)
gather/permute mismatch 3d1b58ba != 507ed7ab (@3)
gather/permute mismatch 238e1f29 != 2eb141f2 (@4)
gather/permute mismatch 19495cff != 327b23c6 (@5)
gather/permute mismatch 507ed7ab != 3d1b58ba (@6)
gather/permute mismatch 2ae8944a != 625558ec (@7)
gather/permute mismatch 74b0dc51 != 41b71efb (@8)
gather/permute mismatch 46e87ccd != 643c9869 (@9)
gather/permute mismatch 79e2a9e3 != 66334873 (@10)
gather/permute mismatch 7545e146 != 6b8b4567 (@11)
gather/permute mismatch 643c9869 != 46e87ccd (@12)
gather/permute mismatch 41b71efb != 74b0dc51 (@13)
gather/permute mismatch 6b8b4567 != 7545e146 (@14)
gather/permute mismatch 66334873 != 79e2a9e3 (@15)
gather/permute mismatch 3352255a != 0216231b (@16)
gather/permute mismatch 1190cde7 != 0ded7263 (@17)
gather/permute mismatch 66ef438d != 109cf92e (@18)
gather/permute mismatch 140e0f76 != 1befd79f (@19)
gather/permute mismatch 0ded7263 != 1190cde7 (@20)
gather/permute mismatch 0216231b != 3352255a (@21)
gather/permute mismatch 1befd79f != 140e0f76 (@22)
gather/permute mismatch 109cf92e != 66ef438d (@23)
gather/permute mismatch 5bd062c2 != 12200854 (@24)
gather/permute mismatch 41a7c4c9 != 1f16e9e8 (@25)
gather/permute mismatch 7fdcc233 != 4db127f8 (@26)
gather/permute mismatch 6b68079a != 515f007c (@27)
gather/permute mismatch 1f16e9e8 != 41a7c4c9 (@28)
gather/permute mismatch 12200854 != 5bd062c2 (@29)
gather/permute mismatch 515f007c != 6b68079a (@30)
gather/permute mismatch 4db127f8 != 7fdcc233 (@31)
```

**Comment by dzaima** [comment]
You want `_mm512_setr_epi32`, not `_mm512_set_epi32`.

**Comment by jlainema** [comment]
for (unsigned i=0; i<10; i++) {
    __m512i dx = _mm512_sub_epi32(swizzles_512[i], _mm512_load_epi32(swizzles[i]));
    if (_mm512_reduce_add_epi32(dx))
      printf("index mismatch for register %d\n", i);
  }

This part of the code verifies the indices in the registers and the memory layout are the same.

**Comment by RKSimon** [comment]
Closing as invalid - feel free to open a new issue if you can reduce the internal issue you're seeing.
